transports/webrtc/src/connection/poll_data_channel.rs

```diff
@@ -0,0 +1,289 @@
+                // buffered writer and `File` do).
+                if fut_is_new {
+                    Poll::Ready(Ok(buf.len()))
+                } else {
+                    // If it's the subsequent poll, it's okay to return `Poll::Pending` as it
+                    // indicates that the `PollDataChannel` is not ready for writing. Only one future
+                    // can be in progress at the time.
+                    Poll::Pending
+                }
+            },
+            Poll::Ready(Err(e)) => {
+                self.write_fut = None;
+                Poll::Ready(Err(webrtc_error_to_io(e)))
+            },
+            Poll::Ready(Ok(n)) => {
+                self.write_fut = None;
+                Poll::Ready(Ok(n))
+            },
+        }
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        match self.write_fut.as_mut() {
+            Some(fut) => match fut.as_mut().poll(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(Err(e)) => {
+                    self.write_fut = None;
+                    Poll::Ready(Err(webrtc_error_to_io(e)))
+                },
+                Poll::Ready(Ok(_)) => {
+                    self.write_fut = None;
+                    Poll::Ready(Ok(()))
+                },
+            },
+            None => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let fut = match self.shutdown_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                let dc = self.data_channel.clone();
+                self.shutdown_fut
+                    .get_or_insert(Box::pin(async move { dc.close().await }))
+            },
+        };
+
+        match fut.as_mut().poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Err(e)) => Poll::Ready(Err(webrtc_error_to_io(e))),
+            Poll::Ready(Ok(_)) => Poll::Ready(Ok(())),
+        }
+    }
+}
+
+impl<'a> Clone for PollDataChannel<'a> {
+    fn clone(&self) -> PollDataChannel<'a> {
+        PollDataChannel::new(self.clone_inner())
+    }
+}
+
+impl fmt::Debug for PollDataChannel<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("PollDataChannel")
+            .field("data_channel", &self.data_channel)
+            .finish()
+    }
+}
+
+impl AsRef<DataChannel> for PollDataChannel<'_> {
+    fn as_ref(&self) -> &DataChannel {
+        &*self.data_channel
+    }
+}
+
+fn webrtc_error_to_io(error: Error) -> io::Error {
+    match error {
+        e @ Error::Sctp(webrtc_sctp::Error::ErrEof) => {
+            io::Error::new(io::ErrorKind::UnexpectedEof, e.to_string())
+        },
+        e @ Error::ErrStreamClosed => {
+            io::Error::new(io::ErrorKind::ConnectionAborted, e.to_string())
+        },
+        e => io::Error::new(io::ErrorKind::Other, e.to_string()),
```

Comment:
You can call `e.to_string()` before the match and avoid the additional naming of `e @` then.